velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/exec/WindowFunction.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::window {
+
+namespace {
+
+enum class ValueType {
+  kFirst,
+  kLast,
+};
+
+template <ValueType TValue, typename T>
+class FirstLastValueFunction : public exec::WindowFunction {
+ public:
+  explicit FirstLastValueFunction(
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool, nullptr) {
+    VELOX_CHECK_NULL(args[0].constantValue);
+    valueIndex_ = args[0].index.value();
```

Comment:
Rename this variable as valueColumn_. Also add a VELOX_CHECK for this argument not to be a constant.

Reply:
Thanks for the suggestion but isn't the NULL check on args[0].constantValue enough?

Reply:
I missed the NULL check on the line before. My bad. Apologies !